x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_summary/privilege_summary_calculator.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Role, RoleKibanaPrivilege } from '../../../../../../../common/model';
+import { isGlobalPrivilegeDefinition } from '../../../privilege_utils';
+import { KibanaPrivileges, PrimaryFeaturePrivilege, SecuredFeature } from '../../../../model';
+import { PrivilegeCollection } from '../../../../model/privilege_collection';
+
+export interface EffectiveFeaturePrivileges {
+  [featureId: string]: {
+    primary?: PrimaryFeaturePrivilege;
+    subFeature: string[];
+    hasNonSupersededSubFeaturePrivileges: boolean;
+  };
+}
+export class PrivilegeSummaryCalculator {
+  constructor(private readonly kibanaPrivileges: KibanaPrivileges, private readonly role: Role) {}
+
+  public getEffectiveFeaturePrivileges(entry: RoleKibanaPrivilege): EffectiveFeaturePrivileges {
+    const assignedPrivileges = this.collectAssignedPrivileges(entry);
+
+    const features = this.kibanaPrivileges.getSecuredFeatures();
+
+    return features.reduce((acc, feature) => {
+      const displayedPrimaryFeaturePrivilege = this.getDisplayedPrimaryFeaturePrivilege(
```

Comment:
I was reading through this code again to see if we similarly had a grey area between `read` or `minimal_all`  and began to wonder whether we could take advantage of the `PrivilegeFormCalculator`. Would it be possible to use `PrivilegeFormCalculator::hasSupersededInheritedPrivileges` and when that returns true, just use the privileges from the global entry?

Reply:
I don't think that'll work the way we want it to. `hasSupersededInheritedPrivileges` returns `true` when there is **any** locally assigned privilege which is superseded, not when **all** locally assigned privileges are superseded:


```ts
const role = {
  kibana: [{
    base: [],
    features: {
       discover: ["all"]
    },
    spaces: ["*"]
  }, {
    base: [],
    features: {
       discover: ["read"],
       visualize: ["all"]
    },
    spaces: ["marketing"]
  }]
}
```

This role would have `hasSupersededInheritedPrivileges` evaluate to `true`, even though there are privileges granted which aren't superseded.

If we were to only use the privileges from the global entry, we would not be showing an accurate picture of the effective privileges in the summary. We could consider writing a version of `hasSupersededInheritedPrivileges` which is scoped to a specific feature...something like `hasSupersededPrimaryFeaturePrivilege`?

What is our desired end state?

### Example 1

```ts
const example1 = {
  kibana: [{
    base: [],
    features: {
       discover: ["minimal_all", "url_create"]
    },
    spaces: ["*"]
  }, {
    base: [],
    features: {
       discover: ["read"]
    },
    spaces: ["marketing"]
  }]
}
```

My initial reaction was to show a primary feature privilege of "read" for the marketing space, with that indicator that privileges have been customized, but "read" doesn't fully convey what's granted here. Showing "all" for the marketing space might paint a better picture, but depending on how the sub-feature privileges are defined, that may not be all that accurate either.

Given our _current_ use of sub-feature privileges, I think showing "all" in this case would be more intuitive than the alternative. _However_, that goes against my earlier stance that `read` should be more permissive than `minimal_all` 😅. Fortunately(?), it already behaves this way, because the summary calculator intentionally does not consult the primary/minimal feature privileges in the same order as the form calculator.


### Example 2

```ts
const example1 = {
  kibana: [{
    base: [],
    features: {
       discover: ["read"]
    },
    spaces: ["*"]
  }, {
    base: [],
    features: {
       discover: ["minimal_all"]
    },
    spaces: ["marketing"]
  }]
}
```

Here again, I think I ultimately expect the "Marketing" space to show a primary feature privilege of "all"

Reply:
> We could consider writing a version of hasSupersededInheritedPrivileges which is scoped to a specific feature...something like hasSupersededPrimaryFeaturePrivilege?

I neglected to realize that `hasSupersededInheritedPrivileges` wasn't scoped to a specific feature, that's a great point.

> Given our current use of sub-feature privileges, I think showing "all" in this case would be more intuitive than the alternative.

I agree. I think showing "All" makes more sense as well, at least given our current use. If the user doesn't have sub-feature privileges which match the "all" privilege, we would then display the indicator that their sub-feature privileges have been customized.

> Here again, I think I ultimately expect the "Marketing" space to show a primary feature privilege of "all"

Agreed. However, as soon as the "read" privilege includes some sub-features which "minimal_all" doesn't, I think we should also show the indicator disclosing that their sub-feature privileges have been customized.

Reply:
> > Here again, I think I ultimately expect the "Marketing" space to show a primary feature privilege of "all"
> 
> Agreed. However, as soon as the "read" privilege includes some sub-features which "minimal_all" doesn't, I think we should also show the indicator disclosing that their sub-feature privileges have been customized.

Perfect, sounds like we're on the same page then!